lib/connection.js

```diff
@@ -48,38 +76,51 @@ Connection.prototype._send = function(data, encoding, callback) {
   this.transport.write(data, encoding, callback);
 };
 
-// Mechanism for collecting data,
-// cut by lower-level protocols
-Connection.prototype._onRawData = function(data) {
-  if (data.length < this._bytesToRead) {
-    this._bytesToRead -= data.length;
-    this._data = Buffer.concat([this._data, data]);
-  } else {
-    const newChunk = data.slice(0, this._bytesToRead);
-    this._data = Buffer.concat([this._data, newChunk]);
-    this.emit('data', this._data);
+Connection.prototype._onHandshake = function(buffer) {
+  const handshake = parser.readHandshake(buffer);
+  this.emit('handshake', handshake);
+  this.on('structure', (buffer) => {
+    const struct = parser.readStruct(buffer);
+    if (struct.structType === 'parcel') this._onParcel(struct);
+    else if (struct.structType === 'chunk') this._onChunk(struct);
+  });
+};
 
-    data = data.slice(this._bytesToRead);
-    this._data = Buffer.alloc(0);
-    this._bytesToRead = 0;
+Connection.prototype._onParcel = function(parcel) {
+  const chunks = this._chunks.filter(
+    chunk => chunk.parcelId === parcel.parcelId
+  );
+  const currentLength = chunks.reduce((acc, cur) => acc += cur.length, 0);
+  Object.assign(parcel, { chunks, currentLength });
 
-    if (data.length > 0) {
-      const structType = data.readUInt8();
-      this._bytesToRead = structType === 0 ? 16 : 12;
-      this._onRawData(data);
-    }
+  if (currentLength === parcel.length) {
+    this._emitStructure(parcel);
+  } else {
+    this._parcels.push(parcel);
   }
 };
 
-Connection.prototype._onHandshake = function(buffer) {
-  const handshake = parser.readHandshake(buffer);
-  this.emit('handshake', handshake);
-  this.on('data', this._onStructure);
+Connection.prototype._onChunk = function(chunk) {
+  const parcel = this._parcels.find(
+    parcel => parcel.parcelId === chunk.parcelId
+  );
+
+  if (parcel) {
+    parcel.chunks.push(chunk);
+    parcel.currentLength += chunk.length;
+
+    if (parcel.currentLength === parcel.length) {
+      this._emitStructure(parcel);
+    }
+  } else {
+    this._chunks.push(chunk);
+  }
 };
 
-Connection.prototype._onStructure = function(buffer) {
-  const struct = parser.readStruct(buffer);
-  this.emit(struct.structType, struct);
+Connection.prototype._emitStructure = function(parcel) {
+  parcel.chunks.sort((cur, next) => cur.id - next.id);
+  const payload = parcel.chunks.reduce((acc, cur) => acc += cur.payload, '');
```

Comment:
I think, we should probably convert payload here, according to the encoding given in parcel struct

Reply:
Maybe parser should do this